Add non_maximum_suppression that supports both CPU and GPU arrays.

yuyu2172 opened the pull request:
Merge after #145.

This PR moves `transforms/bbox/non_maximum_suppression.py` to `links/utils/bbox/non_maximum_suppression.py`.

Here are few changes.

+ It supports both CuPy arrays and NumPy arrays. Supporting CuPy is very critical for training Faster RCNN. Usually, NumPy implementation (not the one in this PR, but Cythonized version) is 10x slower than CuPy implementation. Not only that, the magnitude of speed is quite big taking 200ms per iteration for NumPy  version. 
+ Change interface of `non_maximum_suppression` to not return transformed bounding boxes.
+ Add `score` option.
+ Add Cython dependencies. The Cythonized part is important for performance. Plain Python is order of magnitude slower.

yuyu2172 commented:
@Hakuyume 

Can you check this PR?

yuyu2172 commented:
Updated directory structure and rebased dependencies.

Hakuyume commented:
I'm checking this PR. Please wait.

Hakuyume commented on tests/utils_tests/bbox_tests/test_non_maximum_suppression.py:
```diff
@@ -0,0 +1,141 @@
+
+    @attr.gpu
+    def test_non_maximum_suppression_consistency(self):
+        bbox = _generate_bbox(6000, (600, 800), 32, 512)
```
`_generate_bbox` is used only here. It is better to merge it into `setUp`.

yuyu2172 replied:
I had to use `_generate_bbox` in two places, so I left it there.

Hakuyume replied:
I see.

Hakuyume commented on requirements.txt:
```diff
@@ -1,2 +1,3 @@
 chainer >= 1.21
+cython
```
Please add note to `README.md`.

Hakuyume commented on chainercv/utils/bbox/non_maximum_suppression.py:
```diff
@@ -0,0 +1,202 @@
+#define DIVUP(m,n) ((m) / (n) + ((m) % (n) > 0))
+int const threadsPerBlock = sizeof(unsigned long long) * 8;
+
+__device__ inline float devIoU(float const * const a, float const * const b) {
```
Inconsistency between `IoU` and `jaccard`

Hakuyume commented on chainercv/utils/bbox/non_maximum_suppression.py:
```diff
@@ -0,0 +1,202 @@
+    for i, b in enumerate(bbox):
+        lt = np.maximum(b[:2], bbox[selec, :2])
+        rb = np.minimum(b[2:], bbox[selec, 2:])
+        area = np.prod(rb - lt, axis=1) * (lt < rb).all(axis=1)
```
I think this is more simple.
```
for i in score.argsort()[::-1]:
    lt = np.maximum(bbox[i, :2], bbox[selec, :2])
    rb = np.minimum(bbox[i, 2:], bbox[selec, 2:])
...
return selec.nonzero().astype(np.int32)
```

yuyu2172 replied:
Since `selec.nonzero()` returns indices that are not in order with `score.argsort()`, it can not be written as simple as you wrote.
I think reordering `bbox` before for loop is easier to read.

Hakuyume replied:
I thought this function doesn't need to sort the indices according to their scores. The doc says `An array with indices of bounding boxes that are selected`. If we have to return sorted indices, sorting `bbox` before is good as you said.

Hakuyume commented on chainercv/utils/bbox/non_maximum_suppression.py:
```diff
@@ -0,0 +1,202 @@
+    bounding boxes in the image. The second axis represents attributes of
+    the bounding box. They are :obj:`(x_min, y_min, x_max, y_max)`,
+    where the four attributes are coordinates of the bottom left and the
+    top right vertices.
```
I think `x_min, y_min` is coordinates of the top left. This may be my mistake.

Hakuyume commented on chainercv/utils/bbox/non_maximum_suppression.py:
```diff
@@ -0,0 +1,202 @@
+    Returns:
+        array:
+        An array with indices of bounding boxes that are selected. \
+        The shape of this array is :math:`(R,)` and its dtype is\
```
:math:`(R,)` looks the length of the output is same as that of input. How about `(K,)` or `(R',)`? It is better to note `R' <= R`.

Hakuyume commented on tests/utils_tests/bbox_tests/test_non_maximum_suppression.py:
```diff
@@ -0,0 +1,141 @@
+    def check_non_maximum_suppression(self, bbox, threshold, expect):
+        xp = cuda.get_array_module(bbox)
+        selec = non_maximum_suppression(bbox, threshold)
+        self.assertIsInstance(selec, xp.ndarray)
```
https://github.com/pfnet/chainercv/pull/145/files#r115136375

Hakuyume commented on tests/utils_tests/bbox_tests/test_non_maximum_suppression.py:
```diff
@@ -0,0 +1,141 @@
+        # CuPy does not support argsort
+        order = cuda.to_cpu(score).argsort()[::-1]
+        reordered_selec = non_maximum_suppression(
+            bbox[order], threshold, score=None, limit=limit)
```
It is better to pass `limit=None` and truncate the returned array.

Hakuyume commented on tests/utils_tests/bbox_tests/test_non_maximum_suppression.py:
```diff
@@ -0,0 +1,141 @@
+        selec = non_maximum_suppression(bbox, threshold)
+        self.assertIsInstance(selec, xp.ndarray)
+        np.testing.assert_equal(
+            cuda.to_cpu(selec), np.zeros((0,), dtype=np.int32))
```
Testing `selec.shape` is enough.

yuyu2172 commented:
Thanks.
I found a bug in CPU mode when size of bbox is big and score is given.
I changed a test to catch this case b60ad48 and fixed it 4f8caca.

Other than that, I reflected your comments.

yuyu2172 commented:
Jaccard overlap seems to be not a valid word since I could not find a usage of that word on any of the Google search results.

How about using Intersection over Union (IoU)?
Wikipedia says it is one of the correct word.

> The Jaccard index, also known as Intersection over Union and the Jaccard similarity coefficient (From Wikipedia)

Hakuyume commented:
@yuyu2172 Sorry, `Jaccard overlap` is not a valid word. `Jaccard index` is valid. In the area of object detection, `IoU (Intersection over Union)` is more familiar. I think `IoU` is OK. Of course, we have to keep consistency.

yuyu2172 commented:
I reflected our discussions to the code.

Hakuyume commented on chainercv/utils/bbox/non_maximum_suppression.py:
```diff
@@ -0,0 +1,205 @@
+    Args:
+        bbox (array): Bounding boxes to be transformed. The shape is
+            :math:`(R, 4)`. :math:`R` is the number of bounding boxes.
+        thresh (float): Thresold of IoU.
```
Typo, `threshold`.

Hakuyume commented on chainercv/utils/bbox/non_maximum_suppression.py:
```diff
@@ -0,0 +1,205 @@
+    """Suppress bounding boxes according to their IoUs.
+
+    This method checks each bounding box sequentially and selects the bounding
+    box if the Intersection over Union (IoU) between the bounding box and the
```
IoU**s**? You are using `IoU` as a countable noun in this doc.

yuyu2172 commented:
Fixed.

Hakuyume commented:
LGTM
